src/java.base/share/classes/java/lang/StringTemplate.java

```diff
@@ -0,0 +1,612 @@
+     *          Iterator<String> fragmentsIter = st.fragments().iterator();
+     *
+     *          for (Object value : st.values()) {
+     *              sb.append(fragmentsIter.next());
+     *
+     *              if (value instanceof Boolean) {
+     *                  throw new IllegalArgumentException("I don't like Booleans");
+     *              }
+     *
+     *              sb.append(value);
+     *          }
+     *
+     *          sb.append(fragmentsIter.next());
+     *
+     *          return sb.toString();
+     *     }
+     * }
+     *
+     * MyProcessor myProcessor = new MyProcessor();
+     * try {
+     *     int x = 10;
+     *     int y = 20;
+     *     String result = myProcessor."\{x} + \{y} = \{x + y}";
+     *     ...
+     * } catch (IllegalArgumentException ex) {
+     *     ...
+     * }
+     * }
+     * Implementations of this interface may provide, but are not limited to, validating
+     * inputs, composing inputs into a result, and transforming an intermediate string
+     * result to a non-string value before delivering the final result.
+     * <p>
+     * The user has the option of validating inputs used in composition. For example an SQL
+     * processor could prevent injection vulnerabilities by sanitizing inputs or throwing an
+     * exception of type {@code E} if an SQL statement is a potential vulnerability.
+     * <p>
+     * Composing allows user control over how the result is assembled. Most often, a
+     * user will construct a new string from the string template, with placeholders
+     * replaced by string representations of value list elements. These string
+     * representations are created as if invoking {@link String#valueOf}.
+     * <p>
+     * Transforming allows the processor to return something other than a string. For
+     * instance, a JSON processor could return a JSON object, by parsing the string created
+     * by composition, instead of the composed string.
+     * <p>
+     * {@link Processor} is a {@link FunctionalInterface}. This permits
+     * declaration of a processor using lambda expressions;
+     * {@snippet :
+     * Processor<String, RuntimeException> processor = st -> {
+     *     List<String> fragments = st.fragments();
+     *     List<Object> values = st.values();
+     *     // check or manipulate the fragments and/or values
+     *     ...
+     *     return StringTemplate.interpolate(fragments, values);
+     * };
+     * }
+     * The {@link StringTemplate#interpolate()} method is available for those processors
+     * that just need to work with the string interpolation;
+     * {@snippet :
+     * Processor<String, RuntimeException> processor = StringTemplate::interpolate;
+     * }
+     * or simply transform the string interpolation into something other than
+     * {@link String};
+     * {@snippet :
+     * Processor<JSONObject, RuntimeException> jsonProcessor = st -> new JSONObject(st.interpolate());
```

Comment:
This isn't quite a good example; it defeats the point of string templates, that injection attacks like `}, "another_key": {` can still happen with JSON.